docs: add contribution documentation (#60)
strapi/client: CONTRIBUTING.md

- [📋 Prerequisites](#-prerequisites)
- [🛠 Setting Up the Project](#-setting-up-the-project)
  - [1. Clone the Repository](#1-clone-the-repository)
  - [2. Install Dependencies](#2-install-dependencies)
- [🏗️ Compiling the Changes](#-compiling-the-changes)
  - [Build](#build)
  - [Watch](#watch)
  - [Output](#output)
- [🧪 Testing the Changes](#-testing-the-changes)
  - [Unit Tests](#unit-tests)
    - [Running Unit Tests](#running-unit-tests)
    - [Test Coverage](#test-coverage)
    - [Configuration](#configuration)
    - [Writing Tests](#writing-tests)
    - [Mocks](#mocks)
  - [Demos](#demos)
    - [Setting Up the Demos](#setting-up-the-demos)
    - [Run the Demos](#run-the-demos)
- [🔎 Verifying Your Contribution](#-verifying-your-contribution)
- [📝 Commit Messages](#-commit-messages)
- [🎉 You're Ready to Contribute!](#-youre-ready-to-contribute)
- [🔐 Pull Request Reviews](#-pull-request-reviews)
  - [The Review](#the-review)
  - [Merging a PR](#merging-a-pr)
- [📚 Releases](#-releases)

---

## 📋 Prerequisites

Before you get started, ensure you have the following installed:

- **Node.js**: >=20.18.2 <=22.x.x. You can check your Node.js version using:
  ```bash
  node -v
  ```

> - **pnpm**: the project uses `pnpm` as a package manager.

Reviewer: ```suggestion
- **pnpm**: the project requires `pnpm` as the package manager.
```

Both commands rely on the project's [rollup configuration](rollup.config.mjs).

### Build

Run the following command to build the source code:

```bash
pnpm build
```

### Watch

To automatically rebuild the project when changes are detected:

```bash
pnpm watch
```

Now, any changes you make to the `src` folder will trigger a rebuild.

### Output

The build process outputs the generated artifacts to the `dist` folder. These include:

- **Type Definitions**: TypeScript type definitions (`*.d.ts` files).
- **Bundles**: JavaScript bundles in the following formats:
  - **CommonJS (`cjs`)**: useful for Node.js applications or environments requiring CJS modules.
  - **ES Modules (`mjs`)**: ideal for modern JavaScript, TypeScript, and bundlers like Rollup or Webpack.
  - **Browser (`iife`)**: minified bundle for direct consumption in browser environments (for use in `<script>` tags).

---

## 🧪 Testing the Changes

### Unit Tests

Unit tests ensure the reliability and correctness of the features. The project
uses [Jest](https://jestjs.io) as the testing framework.

The project is configured for at least 95% of coverage on branches, functions, lines, and statement, but the project

> aims to stay at 100% while it is possible.

Reviewer: ```suggestion
aims to stay at 100% as long as possible.
```

```typescript
import { URLValidator } from '../../../src/validators';

/**
 * Class representing a FlakyURLValidator which extends URLValidator.
 *
 * This validator is designed to throw an error unexpectedly upon validation and should only be used in test suites.
 */
export class MockFlakyURLValidator extends URLValidator {
  validate() {
    throw new Error('Unexpected error');
  }
}
```

Feel free to create your own mocks based on your needs and export them from
the [folder's index.ts](tests/unit/mocks/index.ts).

### Demos

The repository includes demo projects to allow you to test your changes in different real-world scenarios.

#### Setting Up the Demos

From the project's root, set up the demo apps

```bash
pnpm demo setup
```

This installs the demos dependencies, builds them if needed, and sets up the Strapi app used by the different demos.

#### Run the Demos

To test your changes in the demo projects, start by running the Strapi app:

```bash
pnpm demo app:start
```

Then navigate to any of the example in the [./demo](./demo) folder, and refer to their respective `package.json` for
instructions on how to run them.

> > **Note:** feel free to locally modify the demos to test your changes but avoid commiting them.

Reviewer: ```suggestion
> **Note:** feel free to locally modify the demos to test your changes but avoid commiting them unless they provide fixes or additional functionality.
```
